File: tools/pipeline/compare-analysis-frame.mjs
// Compare a replay's Analyze and Frame artifacts against the historical run they replay.
//
//   node tools/pipeline/compare-analysis-frame.mjs --replay <run-id> --against <run-id>
//
// Reports the structural facts a human should not have to compute — selected and rejected sources,
// cluster decisions, per-thread allocations and words per source, validation attempts, warnings,
// tokens, cost, and the assembled context manifests — and states plainly which of its conclusions
// are deterministic and which are not. It makes no model call and no editorial judgement: the
// numbers are facts, and the reading is left to the reviewer.

import { readFile, readdir } from "node:fs/promises";
import path from "node:path";
import process from "node:process";

import { ROOT, RUNS_DIRECTORY, option } from "../lib/shared.mjs";
import { ANALYSIS_CLUSTER_KEYS, validateAnalysisSelection, validateFrame } from "./editorial-validation.mjs";
import { STYLE_PROFILES, styleProfileFor } from "./style-profiles.mjs";

const runDir = (id) => path.join(ROOT, RUNS_DIRECTORY, id);
const readJson = async (file) => JSON.parse(await readFile(file, "utf8"));

async function tryJson(file) {
  try {
    return await readJson(file);
  } catch {
    return null;
  }
}

const pad = (value, width) => String(value ?? "").padEnd(width);
const num = (value) => (value === null || value === undefined ? "-" : String(value));

// ---------------------------------------------------------------------------------------
// Reading the two runs
// ---------------------------------------------------------------------------------------

async function readRun(runId) {
  const root = runDir(runId);
  const pipeline = await tryJson(path.join(root, "pipeline.json"));
  const summary = await tryJson(path.join(root, "run-summary.json"));
  const records = await tryJson(path.join(root, "stage-records.json"));
  const analysis = await tryJson(path.join(root, "analyze", "output", "analysis.json"));
  const frame = await tryJson(path.join(root, "frame", "output", "frame.json"));
  const corpus = await tryJson(path.join(root, "source-acquisition", "sources.json"));

  const attempts = {};
  for (const stage of ["analyze", "frame"]) {
    const dir = path.join(root, stage, "attempts");
    const entries = await readdir(dir, { withFileTypes: true }).catch(() => []);
    attempts[stage] = [];
    for (const entry of entries.filter((item) => item.isDirectory() && /^attempt-\d+$/.test(item.name)).sort((a, b) => a.name.localeCompare(b.name))) {
      const attemptDir = path.join(dir, entry.name);
      attempts[stage].push({
        attempt: entry.name,
        attempt_record: await tryJson(path.join(attemptDir, "attempt.json")),
        completed: await tryJson(path.join(attemptDir, "completed.json")),
        validation: await tryJson(path.join(attemptDir, "validation.json")),
        context: await tryJson(path.join(attemptDir, "context-manifest.json")),
        corpus: await tryJson(path.join(attemptDir, "corpus-context.json")),
        error: await readFile(path.join(attemptDir, "stage-error.log"), "utf8").catch(() => null),
      });
    }
  }

  return {
    runId,
    pipeline,
    summary,
    records,
    analysis,
    frame,
    corpus,
    attempts,
    stageRecord: Object.fromEntries((records?.stages ?? []).map((record) => [record.stage, record])),
  };
}

// ---------------------------------------------------------------------------------------
// Reporting
// ---------------------------------------------------------------------------------------

function reportCorpus(run) {
  console.log(`corpus: ${run.corpus?.sources?.length ?? 0} reviewed sources`);
}

function reportAttempts(run) {
  console.log("\n--- attempts and validation (deterministic) ---");
  for (const [stage, attempts] of Object.entries(run.attempts)) {
    if (attempts.length === 0) {
      console.log(`  ${pad(stage, 10)} no attempt directories`);
      continue;
    }
    for (const attempt of attempts) {
      const usage = attempt.completed?.usage ?? {};
      const validation = attempt.validation;
      console.log(
        `  ${pad(stage, 10)} ${pad(attempt.attempt, 10)} ` +
        `miss=${num(attempt.completed?.cache_miss_tokens)} out=${num(usage.completion_tokens)} ` +
        `reasoning=${num(usage.completion_tokens_details?.reasoning_tokens)} ` +
        `seconds=${attempt.attempt_record?.started_at && attempt.completed?.completed_at
          ? ((new Date(attempt.completed.completed_at) - new Date(attempt.attempt_record.started_at)) / 1000).toFixed(1)
          : "-"}` +
        `${attempt.attempt_record?.validation_correction ? "  [CORRECTION ATTEMPT]" : ""}` +
        `${validation ? `  validation: ok=${validation.ok} gate=${validation.counts?.gate} advisory=${validation.counts?.advisory}` : ""}`,
      );
      if (validation && !validation.ok) {
        for (const violation of validation.violations ?? []) {
          console.log(`        gate  [${violation.code}]${violation.unit ? ` (${violation.unit})` : ""} ${violation.message}`);
        }
        for (const warning of validation.warnings ?? []) {
          console.log(`        note  [${warning.code}]${warning.unit ? ` (${warning.unit})` : ""} ${warning.message}`);
        }
      }
      if (attempt.error) console.log(`        stage-error.log present (${attempt.error.split("\n")[0]})`);
    }
  }
}

function reportManifests(run, label) {
  console.log(`\n--- assembled context manifests: ${label} ---`);
  for (const [stage, attempts] of Object.entries(run.attempts)) {
    for (const attempt of attempts) {
      if (!attempt.context) continue;
      console.log(`  ${stage}/${attempt.attempt}:`);
      for (const entry of attempt.context.documents) {
        const sections = entry.sections !== undefined ? ` [${entry.sections.join(" | ")}]` : "";
        console.log(`      ${entry.bytes} B  ${entry.path}${sections}`);
      }
      if (attempt.corpus) {
        console.log(
          `      corpus: policy=${attempt.corpus.effective_policy} sources=${attempt.corpus.source_count} ` +
          `recovery=${attempt.corpus.recovery ?? "none"}`,
        );
      }
    }
  }
}

function reportAnalysis(run, label, profile) {
  console.log(`\n--- analysis decisions: ${label} (deterministic extraction) ---`);
  const analysis = run.analysis;
  if (!analysis) {
    console.log("  no analysis.json");
    return;
  }
  const clusterKey = ANALYSIS_CLUSTER_KEYS.find((key) => Array.isArray(analysis[key])) ?? null;
  const clusters = clusterKey ? analysis[clusterKey] : [];
  const verdict = validateAnalysisSelection({ analysis, profile });
  console.log(`  container key: ${clusterKey ?? "NONE RECOGNIZED"} | clusters: ${clusters.length} | alternatives_considered: ${(analysis.alternatives_considered ?? []).length}`);
  console.log(`  validator: ok=${verdict.ok} gate=${verdict.counts.gate} advisory=${verdict.counts.advisory}${verdict.skipped ? " (skipped: profile enforces nothing)" : ""}${verdict.cluster_key !== undefined ? ` cluster_key=${verdict.cluster_key}` : ""}`);
  for (const violation of verdict.violations ?? []) {
    console.log(`    gate  [${violation.code}]${violation.unit ? ` (${violation.unit})` : ""} ${violation.message}`);
  }
  for (const warning of verdict.warnings ?? []) {
    console.log(`    note  [${warning.code}]${warning.unit ? ` (${warning.unit})` : ""} ${warning.message}`);
  }
  const decisions = {};
  for (const cluster of clusters) {
    const decision = cluster.selection_decision ?? cluster.decision ?? "(none)";
    decisions[decision] = (decisions[decision] ?? 0) + 1;
  }
  console.log(`  decisions: ${JSON.stringify(decisions)}`);
  const relationships = {};
  for (const cluster of clusters) {
    const relationship = cluster.relationship_type ?? "(none)";
    relationships[relationship] = (relationships[relationship] ?? 0) + 1;
  }
  console.log(`  relationship types: ${JSON.stringify(relationships)}`);
  for (const cluster of clusters) {
    const numbers = (cluster.source_numbers ?? []).map(Number);
    console.log(
      `    ${pad(cluster.cluster_id ?? cluster.id, 6)} ${pad(cluster.selection_decision ?? cluster.decision, 12)} ` +
      `${pad(cluster.relationship_type, 26)} n=${pad(numbers.length, 3)} [${numbers.join(", ")}]`,
    );
    console.log(`           ${String(cluster.concrete_subject ?? cluster.title_direction ?? "").slice(0, 150)}`);
    if (cluster.value_basis !== undefined) console.log(`           value_basis: ${String(cluster.value_basis).slice(0, 110)}`);
    const contributions = cluster.source_contributions;
    if (Array.isArray(contributions)) {
      for (const entry of contributions) {
        console.log(`             #${entry.source_number}: ${String(entry.unique_contribution ?? "").slice(0, 130)}`);
      }
    }
  }
  console.log("  alternatives_considered:");
  for (const alternative of analysis.alternatives_considered ?? []) {
    console.log(`    - ${String(alternative.subject ?? alternative.candidate ?? JSON.stringify(alternative)).slice(0, 130)}`);
    if (alternative.reason_demoted || alternative.why_not_selected) {
      console.log(`      ${String(alternative.reason_demoted ?? alternative.why_not_selected).slice(0, 130)}`);
    }
  }
}

function reportFrame(run, label, profile) {
  console.log(`\n--- frame plan: ${label} ---`);
  const frame = run.frame;
  if (!frame) {
    console.log("  no frame.json");
    return;
  }
  const units = frame.editorial_units ?? [];
  const retained = units.filter((unit) => (unit.disposition ?? "keep") === "keep");
  console.log(`  mode: ${frame.mode ?? "(undeclared)"} | units: ${units.length} | retained: ${retained.length}`);
  console.log(`  budget block: ${JSON.stringify(frame.budget?.big_picture_words ?? frame.budget?.big_picture_target_words ?? frame.budget?.opening_orientation_words ?? "-")} opening`);
  for (const unit of units) {
    const numbers = (unit.selected_source_numbers ?? []).map(Number);
    const words = unit.depth_target_words;
    const perSource = typeof words === "number" && numbers.length ? (words / numbers.length).toFixed(1) : "-";
    console.log(
      `    ${pad(unit.unit_id ?? unit.id, 6)} ${pad(unit.disposition ?? "keep", 8)} ` +
      `words=${pad(words, 6)} sources=${pad(numbers.length, 3)} per_source=${pad(perSource, 6)} ` +
      `shape=${unit.explanation_shape ?? "-"}`,
    );
    console.log(`           ${String(unit.working_title ?? "").slice(0, 140)}`);
    console.log(`           [${numbers.join(", ")}]`);
  }
  if (frame.budget) {
    console.log(`  budget: ${JSON.stringify(frame.budget).slice(0, 400)}`);
  }
  const verdict = validateFrame({ frame, corpus: run.corpus, profile });
  console.log(`  validator: ok=${verdict.ok} gate=${verdict.counts.gate} advisory=${verdict.counts.advisory}`);
  for (const violation of verdict.violations) {
    console.log(`    gate  [${violation.code}]${violation.unit ? ` (${violation.unit})` : ""} ${violation.message}`);
  }
  for (const warning of verdict.warnings) {
    console.log(`    note  [${warning.code}]${warning.unit ? ` (${warning.unit})` : ""} ${warning.message}`);
  }
  if (verdict.arithmetic) console.log(`  arithmetic: ${JSON.stringify(verdict.arithmetic)}`);
}

function reportCoverage(replay, baseline) {
  console.log("\n--- narrative coverage of the corpus (deterministic) ---");
  const all = new Set((replay.corpus?.sources ?? []).map((source) => Number(source.source_number)));
  const narrative = new Set();
  for (const unit of replay.frame?.editorial_units ?? []) {
    if ((unit.disposition ?? "keep") !== "keep") continue;
    for (const number of unit.selected_source_numbers ?? []) narrative.add(Number(number));
  }
  const baselineNarrative = new Set();
  for (const unit of baseline.frame?.editorial_units ?? []) {
    if ((unit.disposition ?? "keep") !== "keep") continue;
    for (const number of unit.selected_source_numbers ?? []) baselineNarrative.add(Number(number));
  }
  console.log(`  corpus: ${all.size}`);
  console.log(`  replay narrative: ${narrative.size} -> [${[...narrative].sort((a, b) => a - b).join(", ")}]`);
  console.log(`  baseline narrative: ${baselineNarrative.size} -> [${[...baselineNarrative].sort((a, b) => a - b).join(", ")}]`);
  console.log(`  catalog-only in replay: ${[...all].filter((number) => !narrative.has(number)).length} sources`);
}

function reportCost(replay, baseline) {
  console.log("\n--- cost and duration (deterministic) ---");
  for (const [label, run] of [["baseline", baseline], ["replay", replay]]) {
    const summary = run.summary;
    if (!summary) {
      console.log(`  ${pad(label, 9)} no run-summary.json (a partial run still writes one)`);
      continue;
    }
    console.log(
      `  ${pad(label, 9)} total=$${num(summary.cost_usd?.actual)} seconds=${num(summary.total_seconds)} ` +
      `miss=${num(summary.tokens?.cache_miss)} out=${num(summary.tokens?.output)} reasoning=${num(summary.tokens?.reasoning)}`,
    );
    for (const stage of summary.stages ?? []) {
      console.log(
        `      ${pad(stage.stage, 12)} $${num(stage.cost_usd)} ${pad(stage.seconds + "s", 9)} ` +
        `miss=${pad(stage.cache_miss_tokens, 9)} out=${pad(stage.output_tokens, 8)}` +
        `${stage.attempt_count ? ` attempts=${stage.attempt_count}` : ""}` +
        `${stage.model_seconds !== undefined ? ` model=${stage.model_seconds}s` : ""}`,
      );
    }
  }
}

function reportWarnings(run, label) {
  console.log(`\n--- warnings: ${label} ---`);
  const warnings = run.records?.warnings ?? [];
  console.log(warnings.length ? warnings.map((warning) => `  ${warning}`).join("\n") : "  none");
  const degraded = run.records?.degraded_stages ?? [];
  console.log(`  degraded stages: ${degraded.length ? degraded.join(", ") : "none"}`);
}

// ---------------------------------------------------------------------------------------

const replayId = option("--replay");
const baselineId = option("--against");
if (!replayId || !baselineId) {
  console.error("usage: node tools/pipeline/compare-analysis-frame.mjs --replay <run-id> --against <run-id>");
  process.exitCode = 1;
} else {
  const replay = await readRun(replayId);
  const baseline = await readRun(baselineId);
  const profileId = replay.pipeline?.style_profile_id ?? "synthesis-max-v1";
  const profile = styleProfileFor(profileId) ?? STYLE_PROFILES["synthesis-max-v1"];

  console.log(`replay ${replayId} (profile ${profileId} v${replay.pipeline?.style_profile_version ?? "?"}, partial=${Boolean(replay.pipeline?.partial_run)})`);
  console.log(`against ${baselineId} (profile ${baseline.pipeline?.style_profile_id ?? "n/a"})`);
  reportCorpus(replay);
  reportAttempts(replay);
  reportManifests(replay, "replay");
  reportManifests(baseline, "baseline");
  reportAnalysis(replay, "replay", profile);
  reportAnalysis(baseline, "baseline", profile);
  reportFrame(replay, "replay", profile);
  reportFrame(baseline, "baseline", profile);
  reportCoverage(replay, baseline);
  reportCost(replay, baseline);
  reportWarnings(replay, "replay");
  console.log(
    "\nEverything above is deterministic: extracted from artifacts, measured, or recomputed by the\n" +
    "validator. Whether the selections are better is a reading judgement this script does not make.",
  );
}

File: scripts/compare_analysis_frame.py
#!/usr/bin/env python
"""Compare a replay's Analyze and Frame artifacts against the historical run they replay.

Python port of ``scripts/compare-analysis-frame.mjs``.

    python scripts/compare_analysis_frame.py --replay <run-id> --against <run-id>

Reports the structural facts a human should not have to compute — selected and rejected
sources, cluster decisions, per-thread allocations and words per source, validation attempts,
warnings, tokens, cost, and the assembled context manifests — and states plainly which of its
conclusions are deterministic and which are not. It makes no model call and no editorial
judgement: the numbers are facts, and the reading is left to the reviewer.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from digest_system.config.profiles import STYLE_PROFILES, style_profile_for  # noqa: E402
from digest_system.editorial.validation.editorial import (  # noqa: E402
    ANALYSIS_CLUSTER_KEYS,
    validate_analysis_selection,
    validate_frame,
)
from digest_system.runtime.artifacts import ROOT, RUNS_DIRECTORY  # noqa: E402

from _maintenance import configure_stdio, option, try_json  # noqa: E402


def _pad(value, width: int) -> str:
    return str(value if value is not None else "").ljust(width)


def _num(value) -> str:
    return "-" if value is None else str(value)


def _seconds_between(start, end) -> str:
    from datetime import datetime

    if not start or not end:
        return "-"

    def parse(value: str) -> datetime:
        text = str(value)
        if text.endswith("Z"):
            text = text[:-1] + "+00:00"
        return datetime.fromisoformat(text)

    return f"{(parse(end) - parse(start)).total_seconds():.1f}"


def read_run(run_id: str) -> dict:
    root = ROOT / RUNS_DIRECTORY / run_id
    attempts: dict[str, list] = {}
    for stage in ("analyze", "frame"):
        directory = root / stage / "attempts"
        entries = (
            sorted(
                (entry for entry in directory.iterdir() if entry.is_dir() and entry.name.startswith("attempt-")),
                key=lambda entry: entry.name,
            )
            if directory.is_dir()
            else []
        )
        attempts[stage] = []
        for entry in entries:
            attempt_dir = directory / entry.name
            error = None
            error_path = attempt_dir / "stage-error.log"
            if error_path.exists():
                error = error_path.read_text(encoding="utf-8")
            attempts[stage].append(
                {
                    "attempt": entry.name,
                    "attempt_record": try_json(attempt_dir / "attempt.json"),
                    "completed": try_json(attempt_dir / "completed.json"),
                    "validation": try_json(attempt_dir / "validation.json"),
                    "context": try_json(attempt_dir / "context-manifest.json"),
                    "corpus": try_json(attempt_dir / "corpus-context.json"),
                    "error": error,
                }
            )
    records = try_json(root / "stage-records.json")
    return {
        "runId": run_id,
        "pipeline": try_json(root / "pipeline.json"),
        "summary": try_json(root / "run-summary.json"),
        "records": records,
        "analysis": try_json(root / "analyze" / "output" / "analysis.json"),
        "frame": try_json(root / "frame" / "output" / "frame.json"),
        "corpus": try_json(root / "source-acquisition" / "sources.json"),
        "attempts": attempts,
        "stageRecord": {record["stage"]: record for record in (records or {}).get("stages", [])},
    }


def report_attempts(run: dict) -> None:
    print("\n--- attempts and validation (deterministic) ---")
    for stage, attempts in run["attempts"].items():
        if not attempts:
            print(f"  {_pad(stage, 10)} no attempt directories")
            continue
        for attempt in attempts:
            usage = (attempt["completed"] or {}).get("usage") or {}
            validation = attempt["validation"]
            line = (
                f"  {_pad(stage, 10)} {_pad(attempt['attempt'], 10)} "
                f"miss={_num((attempt['completed'] or {}).get('cache_miss_tokens'))} out={_num(usage.get('completion_tokens'))} "
                f"reasoning={_num((usage.get('completion_tokens_details') or {}).get('reasoning_tokens'))} "
                f"seconds={_seconds_between((attempt['attempt_record'] or {}).get('started_at'), (attempt['completed'] or {}).get('completed_at'))}"
            )
            if (attempt["attempt_record"] or {}).get("validation_correction"):
                line += "  [CORRECTION ATTEMPT]"
            if validation:
                line += f"  validation: ok={validation['ok']} gate={validation['counts']['gate']} advisory={validation['counts']['advisory']}"
            print(line)
            if validation and not validation["ok"]:
                for violation in validation.get("violations") or []:
                    unit = f" ({violation['unit']})" if violation.get("unit") else ""
                    print(f"        gate  [{violation['code']}]{unit} {violation['message']}")
                for warning in validation.get("warnings") or []:
                    unit = f" ({warning['unit']})" if warning.get("unit") else ""
                    print(f"        note  [{warning['code']}]{unit} {warning['message']}")
            if attempt["error"]:
                print(f"        stage-error.log present ({attempt['error'].splitlines()[0]})")


def report_manifests(run: dict, label: str) -> None:
    print(f"\n--- assembled context manifests: {label} ---")
    for stage, attempts in run["attempts"].items():
        for attempt in attempts:
            if not attempt["context"]:
                continue
            print(f"  {stage}/{attempt['attempt']}:")
            for entry in attempt["context"]["documents"]:
                sections = f" [{' | '.join(entry['sections'])}]" if entry.get("sections") is not None else ""
                print(f"      {entry['bytes']} B  {entry['path']}{sections}")
            if attempt["corpus"]:
                print(
                    f"      corpus: policy={attempt['corpus']['effective_policy']} sources={attempt['corpus']['source_count']} "
                    f"recovery={attempt['corpus'].get('recovery') or 'none'}"
                )


def report_analysis(run: dict, label: str, profile) -> None:
    print(f"\n--- analysis decisions: {label} (deterministic extraction) ---")
    analysis = run["analysis"]
    if not analysis:
        print("  no analysis.json")
        return
    cluster_key = next((key for key in ANALYSIS_CLUSTER_KEYS if isinstance(analysis.get(key), list)), None)
    clusters = analysis[cluster_key] if cluster_key else []
    verdict = validate_analysis_selection(analysis=analysis, profile=profile)
    print(
        f"  container key: {cluster_key or 'NONE RECOGNIZED'} | clusters: {len(clusters)} | "
        f"alternatives_considered: {len(analysis.get('alternatives_considered') or [])}"
    )
    print(
        f"  validator: ok={verdict['ok']} gate={verdict['counts']['gate']} advisory={verdict['counts']['advisory']}"
        + (" (skipped: profile enforces nothing)" if verdict.get("skipped") else "")
        + (f" cluster_key={verdict['cluster_key']}" if verdict.get("cluster_key") is not None else "")
    )
    for violation in verdict.get("violations") or []:
        unit = f" ({violation['unit']})" if violation.get("unit") else ""
        print(f"    gate  [{violation['code']}]{unit} {violation['message']}")
    for warning in verdict.get("warnings") or []:
        unit = f" ({warning['unit']})" if warning.get("unit") else ""
        print(f"    note  [{warning['code']}]{unit} {warning['message']}")
    decisions: dict[str, int] = {}
    for cluster in clusters:
        decision = cluster.get("selection_decision", cluster.get("decision", "(none)"))
        decisions[decision] = decisions.get(decision, 0) + 1
    print(f"  decisions: {json.dumps(decisions)}")
    relationships: dict[str, int] = {}
    for cluster in clusters:
        relationship = cluster.get("relationship_type", "(none)")
        relationships[relationship] = relationships.get(relationship, 0) + 1
    print(f"  relationship types: {json.dumps(relationships)}")
    for cluster in clusters:
        numbers = [int(number) for number in cluster.get("source_numbers") or []]
        print(
            f"    {_pad(cluster.get('cluster_id', cluster.get('id')), 6)} "
            f"{_pad(cluster.get('selection_decision', cluster.get('decision')), 12)} "
            f"{_pad(cluster.get('relationship_type'), 26)} n={_pad(len(numbers), 3)} [{', '.join(map(str, numbers))}]"
        )
        print(f"           {str(cluster.get('concrete_subject', cluster.get('title_direction', '')))[:150]}")
        if cluster.get("value_basis") is not None:
            print(f"           value_basis: {str(cluster['value_basis'])[:110]}")
        contributions = cluster.get("source_contributions")
        if isinstance(contributions, list):
            for entry in contributions:
                print(f"             #{entry.get('source_number')}: {str(entry.get('unique_contribution', ''))[:130]}")
    print("  alternatives_considered:")
    for alternative in analysis.get("alternatives_considered") or []:
        print(f"    - {str(alternative.get('subject', alternative.get('candidate', json.dumps(alternative))))[:130]}")
        reason = alternative.get("reason_demoted") or alternative.get("why_not_selected")
        if reason:
            print(f"      {str(reason)[:130]}")


def report_frame(run: dict, label: str, profile) -> None:
    print(f"\n--- frame plan: {label} ---")
    frame = run["frame"]
    if not frame:
        print("  no frame.json")
        return
    units = frame.get("editorial_units") or []
    retained = [unit for unit in units if (unit.get("disposition", "keep")) == "keep"]
    print(f"  mode: {frame.get('mode', '(undeclared)')} | units: {len(units)} | retained: {len(retained)}")
    budget = frame.get("budget") or {}
    opening = budget.get("big_picture_words", budget.get("big_picture_target_words", budget.get("opening_orientation_words", "-")))
    print(f"  budget block: {json.dumps(opening)} opening")
    for unit in units:
        numbers = [int(number) for number in unit.get("selected_source_numbers") or []]
        words = unit.get("depth_target_words")
        per_source = f"{words / len(numbers):.1f}" if isinstance(words, (int, float)) and numbers else "-"
        print(
            f"    {_pad(unit.get('unit_id', unit.get('id')), 6)} {_pad(unit.get('disposition', 'keep'), 8)} "
            f"words={_pad(words, 6)} sources={_pad(len(numbers), 3)} per_source={_pad(per_source, 6)} "
            f"shape={unit.get('explanation_shape', '-')}"
        )
        print(f"           {str(unit.get('working_title', ''))[:140]}")
        print(f"           [{', '.join(map(str, numbers))}]")
    if frame.get("budget"):
        print(f"  budget: {json.dumps(frame['budget'])[:400]}")
    verdict = validate_frame(frame=frame, corpus=run["corpus"], profile=profile)
    print(f"  validator: ok={verdict['ok']} gate={verdict['counts']['gate']} advisory={verdict['counts']['advisory']}")
    for violation in verdict["violations"]:
        unit = f" ({violation['unit']})" if violation.get("unit") else ""
        print(f"    gate  [{violation['code']}]{unit} {violation['message']}")
    for warning in verdict["warnings"]:
        unit = f" ({warning['unit']})" if warning.get("unit") else ""
        print(f"    note  [{warning['code']}]{unit} {warning['message']}")
    if verdict.get("arithmetic"):
        print(f"  arithmetic: {json.dumps(verdict['arithmetic'])}")


def report_coverage(replay: dict, baseline: dict) -> None:
    print("\n--- narrative coverage of the corpus (deterministic) ---")
    all_numbers = {int(source["source_number"]) for source in (replay["corpus"] or {}).get("sources") or []}

    def narrative(run: dict) -> set[int]:
        numbers: set[int] = set()
        for unit in (run["frame"] or {}).get("editorial_units") or []:
            if (unit.get("disposition", "keep")) != "keep":
                continue
            numbers |= {int(number) for number in unit.get("selected_source_numbers") or []}
        return numbers

    replay_narrative = narrative(replay)
    baseline_narrative = narrative(baseline)
    print(f"  corpus: {len(all_numbers)}")
    print(f"  replay narrative: {len(replay_narrative)} -> [{', '.join(map(str, sorted(replay_narrative)))}]")
    print(f"  baseline narrative: {len(baseline_narrative)} -> [{', '.join(map(str, sorted(baseline_narrative)))}]")
    print(f"  catalog-only in replay: {len(all_numbers - replay_narrative)} sources")


def report_cost(replay: dict, baseline: dict) -> None:
    print("\n--- cost and duration (deterministic) ---")
    for label, run in (("baseline", baseline), ("replay", replay)):
        summary = run["summary"]
        if not summary:
            print(f"  {_pad(label, 9)} no run-summary.json (a partial run still writes one)")
            continue
        cost = summary.get("cost_usd") or {}
        tokens = summary.get("tokens") or {}
        print(
            f"  {_pad(label, 9)} total=${_num(cost.get('actual'))} seconds={_num(summary.get('total_seconds'))} "
            f"miss={_num(tokens.get('cache_miss'))} out={_num(tokens.get('output'))} reasoning={_num(tokens.get('reasoning'))}"
        )
        for stage in summary.get("stages") or []:
            line = (
                f"      {_pad(stage['stage'], 12)} ${_num(stage.get('cost_usd'))} {_pad(str(stage.get('seconds')) + 's', 9)} "
                f"miss={_pad(stage.get('cache_miss_tokens'), 9)} out={_pad(stage.get('output_tokens'), 8)}"
            )
            if stage.get("attempt_count"):
                line += f" attempts={stage['attempt_count']}"
            if stage.get("model_seconds") is not None:
                line += f" model={stage['model_seconds']}s"
            print(line)


def report_warnings(run: dict, label: str) -> None:
    print(f"\n--- warnings: {label} ---")
    warnings = (run["records"] or {}).get("warnings") or []
    print("\n".join(f"  {warning}" for warning in warnings) if warnings else "  none")
    degraded = (run["records"] or {}).get("degraded_stages") or []
    print(f"  degraded stages: {', '.join(degraded) if degraded else 'none'}")


configure_stdio()


def main(argv: list[str] | None = None) -> int:
    argv = list(argv if argv is not None else sys.argv[1:])
    replay_id = option(argv, "--replay")
    baseline_id = option(argv, "--against")
    if not replay_id or not baseline_id:
        print("usage: python scripts/compare_analysis_frame.py --replay <run-id> --against <run-id>", file=sys.stderr)
        return 1

    replay = read_run(replay_id)
    baseline = read_run(baseline_id)
    profile_id = (replay["pipeline"] or {}).get("style_profile_id", "synthesis-max-v1")
    profile = style_profile_for(profile_id) or STYLE_PROFILES["synthesis-max-v1"]

    print(
        f"replay {replay_id} (profile {profile_id} v{(replay['pipeline'] or {}).get('style_profile_version', '?')}, "
        f"partial={bool((replay['pipeline'] or {}).get('partial_run'))})"
    )
    print(f"against {baseline_id} (profile {(baseline['pipeline'] or {}).get('style_profile_id', 'n/a')})")
    print(f"corpus: {len((replay['corpus'] or {}).get('sources') or [])} reviewed sources")
    report_attempts(replay)
    report_manifests(replay, "replay")
    report_manifests(baseline, "baseline")
    report_analysis(replay, "replay", profile)
    report_analysis(baseline, "baseline", profile)
    report_frame(replay, "replay", profile)
    report_frame(baseline, "baseline", profile)
    report_coverage(replay, baseline)
    report_cost(replay, baseline)
    report_warnings(replay, "replay")
    print(
        "\nEverything above is deterministic: extracted from artifacts, measured, or recomputed by the\n"
        "validator. Whether the selections are better is a reading judgement this script does not make."
    )
    return 0


if __name__ == "__main__":  # pragma: no cover - script entry point
    raise SystemExit(main())